### DISCUSSION TITLE
Phases of event loop "call pending callbacks" and "poll for I/O" are unclear in node.js

### DISCUSSION THREAD

**Discussion by dSalieri** [libuv/libuv#4466] [Q&A]
I have looked for information on these two phases wherever I can and nowhere have I been able to find any consensus on the understanding of these phases. Of course, I'm talking primarily about the node.js platform.

_There are no complaints about the description of phases in libuv; it is a low-level library, it is not tied to anything in itself, but node.js is tied to libuv, but it is not going to explain all the phases properly (this is shown by countless questions that have no answer) ._

In the node.js specification, the **call pending callbacks** and **poll for I/O** phases are called: **pending callbacks** (in older documentation this phase was called **callbacks I/O**) and **poll**.

As the node.js documentation describes these phases at the moment:
> **pending callbacks:** executes I/O callbacks deferred to the next loop iteration.
> **poll:** retrieve new I/O events; execute I/O related callbacks (almost all with the exception of close callbacks, the ones scheduled by timers, and setImmediate()); node will block here when appropriate.

As the node.js documentation once described (2017y):
> **I/O callbacks:** executes almost all callbacks with the exception of close callbacks, the ones scheduled by timers, and setImmediate().
> **poll:** retrieve new I/O events; node will block here when appropriate.

---

And reading these two versions, a dissonance occurs: previously callbacks were executed in **pending callbacks** (previously called **I/O callbacks**) and the **poll** phase was only engaged in polling, now the **poll** phase has taken over responsibilities and executes callbacks .

What then does **pending callbacks** do at the moment?

It is said that some I/O callbacks are deferred to the next iteration, but how do you know what exactly is deferred or what events and factors influence the transfer of execution of I/O callbacks to the **pending callbacks** phase from the **poll** phase?

The node.js documentation gives an example with TCP errors. Is this the only case or is there something else?
> It needs examples for each of the **pending callbacks** and **poll** phases. It would also be nice to know specifically what libuv affects the transfer of callbacks from the **poll** phase to the **pending callbacks** phase.

I looked through a bunch of sources, there are no examples with **pending callbacks** anywhere, I even asked this question on SO with a certain frequency - there was no answer.

@bnoordhuis well it's question again to you, because I couldn’t get the answer anywhere. Sorry for asking here :(

**Comment by IsraelAfangideh** [comment]
I am also interested in the answer
